### ISSUE TITLE
[2.15.0] Call to a member function removeQueryCacheProfile() on null

### ISSUE THREAD

**Issue by PhilETaylor** [doctrine/orm#10673]
### BC Break Report

|    Q        |   A
|------------ | ------
| BC Break    | yes
| Version     | 2.15.0

#### Summary

Previously fully working live Symfony 6.3.x-dev/6.2 application working fine, until upgrade to doctrine/orm 2.15.0 this morning. 

#### Previous behavior

The following code worked without issue in a standard Symfony repository 

```php
  public function getChangeLogSinceLastLogin($date): array
    {
        $query = $this->createQueryBuilder('o')
            ->orderBy('o.posted', Criteria::DESC)
            ->where('o.published = 1')
            ->andWhere('o.posted > :date and o.posted < :now')
            ->setParameter('date', $date)
            ->setParameter('now', date('Y-m-d H:i:s'))
            ->getQuery();

        return $query->getResult() ?? [];
    }
```

#### Current behavior

After updating to 2.15.0 the `getResult` method call fatal errors with 

`Call to a member function removeQueryCacheProfile() on null`

<img width="1210" alt="ScreenShot-2023-05-04-12 11 51" src="https://user-images.githubusercontent.com/400092/236187954-0be85965-3999-4883-97b4-f6b212a4a58a.png">

#### How to reproduce

I can work on seeing if I can provide a reproducer after lunch, but posting here in case its a known issue?

**Comment by PhilETaylor** [comment]
Nevermind, removing the cache drivers from the symfony configuration seems to fix this... 

<img width="895" alt="ScreenShot-2023-05-04-12 20 08" src="https://user-images.githubusercontent.com/400092/236189639-c890857a-d088-4374-ac13-ed41c1818cdf.png">

**Comment by derrabus** [comment]
But those settings should be fine. I'd still be interested in a reproducer.

**Comment by PhilETaylor** [comment]
I cannot reproduce it in a clean reproducer app :(

**Comment by PhilETaylor** [comment]
pin pointed it to a configuration of default_redis_provider in my app (https://symfony.com/doc/current/reference/configuration/framework.html#default-redis-provider)

```php
$cache->defaultRedisProvider(env('REDIS_URL'))
```

if I remove this then it works as expected, trying to replicate that in a reproducer now

**Comment by PhilETaylor** [comment]
Looking on my production redis server (that has not yet been deployed this new version of orm to) I can see the underscore property cached... so when this is unserialised, that is what is causing the problem. 

<img width="940" alt="ScreenShot-2023-05-04-14 23 49" src="https://user-images.githubusercontent.com/400092/236217988-37b03fda-c175-4504-b611-089da81a415a.png">

**Comment by derrabus** [comment]
Okay, so it's a deserialization issue. Serialize a parsed ORM query with 2.14 and try to deserialize it with 2.15 and it bombs, did I get this right?

**Comment by PhilETaylor** [comment]
Correct. 

So I guess the reason I could not reproduce it, is that the new version creates the correct cache which is then correctly deserialised. 

I guess (as Im out of time and have a dentist appointment to run to) the reproducer needs to generate a cache with the old version of orm, and then you need to upgrade to the latest version, and have the latest version try to deserialise the cached ParserResult, which will then set the wrong property (the underscored one) and thus leaving the correct property (which no longer has an underscore in the latest version) null.

**Comment by flack** [comment]
Yes, was the same in my case: I use Apcu, and as soon as I restarted Apache (which clears the apcu cache) the error disappeared

**Comment by fezfez** [comment]
I got the same issue with mezzio using acpu cache

**Comment by curry684** [comment]
Can confirm upgrade issue, I'm using the Redis adapter for the query cache. Manually invoking `flushdb` on the relevant Redis DB fixed the issue.

**Comment by greg0ire** [comment]
Looking at the release notes, nothing stands out.
Possible next steps:
- try while upgrading `doctrine/orm` and only that. version 2.15.0 allows upgrading `doctrine/instantiator`, and maybe more packages if that has rippling effects
- try using `git bisect` to pinpoint what commit exactly is responsible for the issue.
- figure out why/how it works with 2.14: does it actually empty Redis, or is there a new namespace created (apparently that's how `doctrine/cache` used to work). Also, it would be great to figure out if you are using `doctrine/cache` or `psr/cache`, and if that somehow changes while upgrading.

**Comment by PhilETaylor** [comment]
> Looking at the release notes, nothing stands out.

Apart from the fact that the private properties were changed from having underscores in them, so when the cached data is unserialised, it attempts to set the underscored private properties, and they dont exist,hence the logged deprecations in my screenshot above. 

This is the root problem here.

**Comment by flack** [comment]
would be nice if the serialized objects had some sort of version number, then the cached data could be ignored by newer releases

**Comment by curry684** [comment]
> would be nice if the serialized objects had some sort of version number, then the cached data could be ignored by newer releases

That's pretty much impossible as you'd need to deserialize the object to get to the version number. There's a reason default Composer config automatically runs `cache:clear` after install/update, because caches are in general not expected or required to be version-compatible.

**Comment by flack** [comment]
or a cache namespace, then

**Comment by greg0ire** [comment]
Yes. Flushing seems sub-optimal since it would mean a performance loss during a deploy. And indeed, it could have happened with any previous release. I would not say the root issue is that underscores were removed, I'd say the root cause is expecting the cache to always stay backwards-compatible, and not bumping a version number in the namespace (if there is still one).

Possible next steps:
- figure out what tools `doctrine/orm` exposes to deal with this. Apparently, you can delete, but you can also invalidate: https://github.com/doctrine/orm/blob/60c625be17668f905cd2c1f73b48ab9b7a105755/lib/Doctrine/ORM/Tools/Console/Command/ClearCache/MetadataCommand.php#L27
- figure out how these tools are integrated in the bundles
- if there is an integration, figure out why it does not kick in.

**Comment by curry684** [comment]
> Flushing seems sub-optimal since it would mean a performance loss during a deploy. 

If the caches are possibly incompatible that's unavoidable anyway. The only way to prevent the performance hit is to prewarm the caches and ship them with the deployment. Symfony has full support for this (you can ship applications with a precompiled container and warmed *filesystem* cache) but it's pretty much moot point as for those application that do that AND depend on filesystem caches the Doctrine caches will by definition be ABSENT during deployment. They always take the performance hit anyway.

I think the root issue is just that the Doctrine Bundle should automatically invoke those 3 cache clear commands during Symfony built in cache clear. It will out of the box solve all portability issues in exchange for sometimes innecessarily rebuilding a cache, and the only way to prevent that is to implement a cache warmer which would be incredibly complex as it would need to be compatible with Redis, APCU, Memcached et al adapters.

**Comment by PhilETaylor** [comment]
For me, understanding what the problem is, and now knowing that once I deploy after upgrading doctrine/orm I need to address the caching issue, makes this whole issue moot. It was the fact that this was not known, and not documented in the release notes that forced me to raise this as an issue while investigating it. 

The issue only affects those people that have existing caches, and who upgrade without clearing those caches manually somehow. Like you said, caches are not part of the b\c promises. 

Knowing this, I would say there is nothing to "resolve" here, other than having the knowledge that this update will break if there are existing caches. 

Maybe we should just close this then as there is not really anything to fix?

**Comment by curry684** [comment]
> Maybe we should just close this then as there is not really anything to fix?

The bug just moved. There is no issue whatsoever with the 2.15.0 release, the fact that the issue popped up now is indeed just an accidental symptom of a perfectly fine B/C change.

There's still a bug though that causes this, because as you say:

> The issue only affects those people that have existing caches, and who upgrade without clearing those caches manually somehow. 

I always run `bin/console cache:clear` between deployments. Manually or automatically. But DoctrineBundle is (apparently) not respecting that. That's the actual root cause bug that causes these deployment issues.

> other than having the knowledge that this update will break if there are existing caches.

You do realize that this is the 99.9% use case of deployments: overwriting an existing older codebase with shiny new stuff?  😉
